File: week3/java_week3/Car/Car.java
public class Car {
    private double fuelAmount;
    private double fuelCapacity;
    private double fuelConsumption;
    private Point location;
    private String model;

    public Car(double capacity, double consumption, Point location, String model) {
        this.fuelAmount = 0;
        this.fuelCapacity = capacity;
        this.fuelConsumption = consumption;
        this.location = location;
        this.model = model;
    }

    public Car() {
        this(60, 0.6, new Point(0,0), "Mercedes");
    }

    public double getFuelAmount() {
        return this.fuelAmount;
    }

    public double getFuelCapacity() {
        return this.fuelCapacity;
    }

    public double getFuelConsumption() {
        return this.fuelConsumption;
    }

    public Point getLocation() {
        return this.location;
    }

    public String getModel() {
        return this.model;
    }

    public void drive(Point destination) throws OutOfFuelException {
        double neededFuel;

        if ( this.fuelConsumption <= 0 ) {
            System.out.println("Incorrect fuel consumption data.");
            return;
        }

        neededFuel = this.location.distance(destination) * this.fuelConsumption;
        if ( this.fuelAmount <= 0 ) {
            throw new OutOfFuelException();
        }

        if ( this.fuelAmount < neededFuel ) {
            throw new OutOfFuelException();
        }

        this.fuelAmount -= neededFuel;
        this.location = destination;
    }

    public void drive(double x, double y) throws OutOfFuelException {
        this.drive(new Point(x, y));
    }

    public void refill(double fuel) throws ToMuchFuelException {
        double newFuelAmount;

        if ( fuel <= 0 ) {
            System.out.println("Incorrect fuel refill data.");
            return;
        }
        
        newFuelAmount = this.fuelAmount + fuel;

        if ( fuel > this.fuelCapacity || newFuelAmount > this.fuelCapacity ) {
            throw new ToMuchFuelException();
        }

        this.fuelAmount = newFuelAmount;
    }

    public void showInfo() {
        System.out.println("Car info\n" 
                        + "===========\n"
                        + "Model           : " + this.getModel() + "\n"
                        + "Capacity        : " + this.getFuelCapacity() + "\n"
                        + "Consumption     : " + this.getFuelConsumption() + "\n"
                        + "Fuel amount     : " + this.getFuelAmount() + "\n"
                        + "Current location: " + this.getLocation());
    }
}

File: week3/cpp_hw_week3/Car/Car.cpp
#include "Car.h"

Car::Car(double capacity, double consumption, const Point& location, const std::string& model) {
    this->fuelAmount = 0;
    this->fuelCapacity = capacity;
    this->fuelConsumption = consumption;
    this->location = location;
    this->model = model;
}

Car::~Car() {

}

double Car::getFuelAmount() const {
    return this->fuelAmount;
}

double Car::getFuelCapacity() const {
    return this->fuelCapacity;
}

double Car::getFuelConsumption() const {
    return this->fuelConsumption;
}

const Point& Car::getLocation() const {
    return this->location;
}

const std::string& Car:: getModel() const {
    return this->model;
}

void Car::drive(const Point& destination) {
    double neededFuel;

    if ( this->fuelConsumption <= 0 ) {
        std::cout << "Incorrect fuel consumption data." << std::endl;
        return;
    }

    neededFuel = this->location.distance(destination) * this->fuelConsumption;
    
    if ( this-> fuelAmount <= 0 ) {
        throw OutOfFuel();
    }

    if ( this->fuelAmount < neededFuel ) {
        std::cout << "Sorry, but you have not enough fuel to get to the destination." 
                  << "Plz refill your tank.\n"
                  << "Current fuel: " << this->fuelAmount << "\n"
                  << "Fuel needed: " << neededFuel << std::endl; 
        throw OutOfFuel();
    }

    this->fuelAmount -= neededFuel;
    this->location = destination;

}

void Car::drive(double x, double y) {
    this->drive(Point(x, y));
}

void Car::refill(double fuel) {
    double newFuelAmount;

    if  ( fuel <= 0 ) {
        std::cout << "Inccorect fuel refill data." << std::endl;
        return;
    }

    newFuelAmount = this->fuelAmount + fuel;
    if ( fuel > this->fuelCapacity || newFuelAmount > this->fuelCapacity ) {
        throw ToMuchFuel();
    }

    this->fuelAmount = newFuelAmount;
}


std::ostream& operator<<(std::ostream& out, const Car& car) {
    out << "Car info" << "\n"
        << "=================" << "\n"
        << "Model            : " << car.getModel() << "\n"
        << "Capacity         : " << car.getFuelCapacity() << "l." <<"\n"
        << "Consumption      : " << car.getFuelConsumption() << "l." <<"\n"
        << "Fuel amount      : " << car.getFuelAmount() <<  "l." << "\n"
        << "Current location : " << car.getLocation();

        return out;
}